Add support for compiling CUDA using Clang

psalz opened the pull request:
This adds support for compiling hipSYCL on CUDA using Clang, as discussed in #22.

This turned out to be easier than expected, as no code changes whatsoever are required for libhipSYCL (even before it was only adjusting the compiler detection macros, but since hipSYCL now uses `__CUDACC__`, which is also defined when compiling with Clang, it works right out of the box). All changes are thus contained to `syclcc`.

Clang CUDA is now the default method for compiling CUDA, the previous behavior (= using nvcc) can be restored by explicitly passing `--hipsycl-platform=nvcc`.

The version of Clang used for compilation can be configured using either the `--cuda-clang-compiler=<compiler>` CLI parameter, or the `HIPSYCL_CUDA_CLANG_COMPILER` environment variable.

In addition, this includes some minor refactorings of some parts of `syclcc` and adds a new configuration option `platform_needs_template_pruning`, which dictates whether the `--with-template-pruning` flag is passed to hipsycl_transform_source. This is only enabled for nvcc.

---

Now I don't think this PR is ready for prime-time just yet, as there are some remaining issues / open questions:
- I wasn't sure if we should compile a separate version of libhipSYCL for clang/nvcc, but my hunch is that it shouldn't be necessary.
- Clang requires a `--cuda-gpu-arch` parameter to be set, as it otherwise defaults to compute capability 2.0, which is [pretty old](https://developer.nvidia.com/cuda-gpus) and only supported by CUDA 7 and 8. I decided to go with `sm_52`, which corresponds to the GTX 9xx series. These should be forward-compatible however. Of course this can be overwritten by passing the parameter to `syclcc`.
- [It looks like](https://releases.llvm.org/7.0.0/docs/CompileCudaWithLLVM.html) Clang 6 and 7 can only target CUDA 7 and 8. I have CUDA 10 installed so I cannot test if everything works with Clang 6/7. Right now the version of Clang used by `syclcc` defaults to `clang++`, which I assume on many systems will be Clang 6 (at least that's the default on Ubuntu 18.04 LTS). Maybe this should be changed to Clang 8 once it releases in February?

illuhad commented:
Excellent, good to hear that things are easier than expected for a change :)

As far as your questions are concerned:
* I'm pretty sure that a separate libhipSYCL for clang/nvcc is not needed. libhipSYCL only interfaces with the CUDA Runtime API. libhipSYCL is only compiled with nvcc because the HIP headers require that they are compiled with HCC or a CUDA compiler. I'm pretty sure that clang also just uses the nvidia runtime API (it's still a proprietary platform, so there are not so many choices..). So, let's first try without a separate libhipSYCL. If we hit problems we can revisit that question ;)
* I'm fine with `sm_52`. Some clusters still have some legacy Kepler GPUs in them, but this is not going to stay for long ;) However, I think that this behavior should be documented in the readme and potentially `syclcc --help`
* In `syclcc` there is already a list of clang executables that will be tried to see if they are available when looking for the CPU backend compiler [here](https://github.com/illuhad/hipSYCL/blob/8f080b64bed9d5210334fae90e76d5b4449c445a/bin/syclcc#L67). With a similar logic, we could add a function `get_newest_clang()` which iterates through a list `['clang++-8', 'clang++-7', 'clang++-6', 'clang++']` and returns the first working command? In that way, the newest clang is always preferred.

illuhad commented:
No need to hurry, I also need some additional time to test this since clang nightly seems to conflict with the regular clang package on Arch Linux, so I'll try installing it in a container as a next step..

My impression was that there is always both `clang++-<major>.<minor>` and `clang++-<major>`. Is that not the case on Ubuntu?

Okay, then I'll update the README in one go, including an updated image for the compilation workflow :)

psalz commented:
That's what's confusing me, I only have either `<major>.<minor>`, or `<major>`, not both. But now that I'm thinking about it, this might just be the nightly package (maybe this is not built by the Ubuntu maintainers).

In any case, I just went for checking all possible variants between 8 and 6. I only do this if no version has been set explicitly, to avoid the overhead.

PS: Your new Travis config seems to always try to pull from this repository, but the PR comes from my fork. I think you might have to construct the remote URL yourself from `TRAVIS_PULL_REQUEST_SLUG`.

illuhad commented on bin/syclcc:
```diff
@@ -222,8 +226,62 @@ class source_file_processor:
+        else:
+          transformed_args.append(arg)
+
+      compiler_args = ["--cuda-gpu-arch=sm_52",
```
Won't this lead to clashes if the user wants to override the the arch and passes `--cuda-gpu-arch` himself? The `hip_compiler` only appends the GPU arch argument based on the content of  `HIPSYCL_GPU_ARCH`, if this environment variable is defined. Perhaps we can have a similar behavior here for the sake of consistency and only add `--cuda-gpu-arch=sm_52` if we don't have an arch defined in `HIPSYCL_GPU_ARCH`?

psalz replied:
Well the user-provided arguments are appended afterwards, so I thought they'd just overwrite the default. *But* I just realized that apparently [it is possible to pass multiple architectures at once](http://llvm.org/docs/CompileCudaWithLLVM.html#invoking-clang), and compile for all of them. So yes, this is not a good solution. Also I didn't realize you already had a `HIPSYCL_GPU_ARCH` in place, as `nvcc` doesn't use it. I will pull this out of `hip_compiler` into a general configuration option, is that okay with you?

illuhad replied:
Perfect, that sounds like a good solution :)

illuhad commented on bin/syclcc:
```diff
@@ -222,8 +226,62 @@ class source_file_processor:
+      if self.includes_linking_stage(args):
+        compiler_args += ["-lcudart_static",
+                           "-ldl",
+                           "-lrt"]
```
We should also conditionally link against libhipSYCL as you propose, since this is the behavior of the other compiler classes in `syclcc`.

illuhad replied:
Sorry, forget that, this is of course handled by `config.hipsycl_common_arguments` :)

psalz replied:
I thought maybe we could at some point extract some of the common functionality that all compilers share into a base class which could then also handle stuff like this.

illuhad replied:
Yes, that makes sense. The current approach was implemented when things were still very simple with only two backends around ;)

illuhad commented on bin/syclcc:
```diff
@@ -447,19 +501,33 @@ class syclcc_config:
+    if self._cuda_clang_compiler == None:
+      self._cuda_clang_compiler = get_first_working_command([
+      "clang++-8.0","clang++-8","clang++-7.0","clang++-7",
+      "clang++-6.0","clang++-6","clang++"])
```
Can we reuse this list in `get_host_compiler()`? Then we only have to maintain one list of clang executable names ;)

psalz replied:
Sure, however would you still like to default to `g++`?

illuhad replied:
`get_host_compiler()` is only used for the CPU backend (which does almost not depend on the source-to-source transformation at all), so there's no inherent advantage to preferring clang. And since g++ comes preinstalled basically everywhere, I think it makes sense to use it as default. In the future we could however perhaps add a configuration flag for this.

illuhad commented on bin/syclcc:
```diff
@@ -447,19 +501,33 @@ class syclcc_config:
+    if len(parsed_args) != 2:
+      raise RuntimeError("Invalid argument: "+arg)
+    return parsed_args[1]
+
```
That function is a very useful addition, I like it :)

psalz commented:
I consolidated the lists of Clang executables between the hipCPU and CUDA-Clang compilers, and added support for `HIPSYCL_GPU_ARCH` to the latter, as discussed. I also rebased everything onto the current master.

I thought some time about how to handle the precedence of `--cuda-gpu-arch` parameters and the `HIPSYCL_GPU_ARCH` environment variable. Since it is possible to pass multiple architectures to Clang, I came to the conclusion that the best solution would be to simply merge them. Only if neither is provided the default (`sm_52`) is used.

I also briefly looked into supporting `HIPSYCL_GPU_ARCH` for the `nvcc` compiler, but the supported architecture parameters are a bit of a mess and I'm not sure how to best handle them. I added a TODO note for that.

---
On a somewhat unrelated note: While I tried the changes to `get_host_compiler()` I noticed that for me the hipCPU backend doesn't compile. I didn't open a separate issue because I don't have time to investigate this thoroughly right now; however I think that it's due to the order of the linker flags that are passed to `g++` (with `clang++` it does compile by the way). GCC requires that all libraries are linked after the object file that needs them. While not strictly required for shared libraries (such as libhipSYCL_cpu), this can still be enforced with the `--as-needed` flag. According to [this](https://stackoverflow.com/a/409470/2095767) the existence of this flag varies between distros, and apparently Arch doesn't specify it (which might explain why you didn't get an error).

illuhad commented:
Thanks, looks good :) I agree that merging `--cuda-gpu-arch` and `HIPSYCL_GPU_ARCH` is a good solution for now.

Travis (which now pulls correctly from your fork :) ) now complains that clang cannot find `libdevice` for `sm_52`. This is probably because clang 6 (as used in the travis build) is incompatible with the CUDA 10 in the docker image, so we should update the CUDA dockerfile to also install clang nightly.

Not sure what the best approach is to fix the CPU backend for `g++`, apart of course from just using `clang++`. I will look into that this evening.

psalz commented:
Yes, the Travis error is most likely due to the Clang version!

As for GCC: I think you can simply swap the order of `hipsycl_common_arguments` and `transformed_args` being added to `compiler_args`. Then it works for me (I still get a warning about `-Wno-unused-command-line-argument` being an unrecognized option though!). Also does the CPU backend really need the rewrite/transform passes? Aren't the `__host__` and `__device__` attributes defined to empty anyways?

illuhad commented:
The CPU backend does need the transformation passes due to one tiny issue (not related to `__device__` or `__host__`): In parallel hierarchical for invocations, all declarations that are not of type `cl::sycl::private_memory<T>` need to go into local memory (`__shared__` memory in CUDA/HIP). These `__shared__` attributes are also inserted during the transformation pass. This is also necessary on the CPU backend as it alters semantics so that this variable should be shared across all parallel OpenMP threads in a work group. This is at the moment implemented with `#define __shared__ static`.

Also it was my plan to use the transformation pass in the future to have some significant performance increase on CPUs: If we detect that no barrier is called we can have the OpenMP parallelization span across work groups instead of work items within a group (and iterate over work items within a group sequentially). Better vectorization, less false sharing and so on.
